### ISSUE TITLE
Array order (C/Fortran) is sometimes misinterpreted when using mmapped arrays

### ISSUE THREAD

**Issue by jzwinck** [numba/numba#2789]
I have this code:
```
@numba.njit(nogil=True)
def std_2d_mask(array, mean, mask):
    """Input array is 2D, plus column-wise mean (1D) and mask (1D).
    Returns column-wise standard deviation.
    Same as array[mask].std(0), but ~10x faster.
    """
    assert array.ndim == 2
    length, width = array.shape
    assert length > 0 and len(mask) == length

    count = 0
    variance = np.zeros(width)  # result is always float

    for row in range(length):
        if not mask[row]:
            continue
        count += 1
        for col in range(width):
            diff = array[row, col] - mean[col]
            variance[col] += diff * diff

    return np.sqrt(variance / count)
```

This mostly works, but I have one case where there appears to be non-deterministic misinterpretation of Fortran vs C ordering of the input array.  I have only noticed it when the array is loaded using `np.load(mmap_mode='r')`.  Unfortunately the array that reproduces the problem is huge, so I cannot post it online, but the dtypes and shapes in the mmapped "npy" files are:

    array: 'descr': '<f8', 'fortran_order': True, 'shape': (143731939, 76)
    mean: 'descr': '<f8', 'fortran_order': False, 'shape': (76,)
    mask: 'descr': '|b1', 'fortran_order': False, 'shape': (143731939,)

The program is multithreaded (hence `nogil=True`).  When it gives the wrong result, it's as if the input array were transposed (i.e. as if Numba thinks it is `order='C'`, but it is not).

I have `order='F'` mostly because that's the default in Pandas (i.e. what you usually get from `DataFrame.values`).

I do not know where the bug lies, but I would appreciate advice on how to diagnose and perhaps solve it.

The problem exists with Numba 0.34, 0.35, and 0.37, using NumPy 1.13.1, Python 3.5, on 64-bit Linux.

**Comment by stuartarchibald** [comment]
This is a guess...
Given the program is multi-threaded (I assume `threads` and not `processes`), I would take a look at which function signatures get compiled first . It may be that a signature that is quite broad and will accept (with promotion) a lot of narrower types is compiled ahead of other more specific versions, in which case it will be used! This may be the root cause of the non-deterministic behaviour. If you look at `std_2d_mask.signatures` it will show the order in which they were compiled. Numba is aware of C/F order and it is captured in the signature at compile time.

I would also check that the `np.read` is producing precisely what is expected, including what happens when a slice is made and how Numba then interprets this input.

Hope this helps.

**Comment by jzwinck** [comment]
The problem persists in Numba 0.39.

**Comment by stuartarchibald** [comment]
Thanks. Did you try out any of the suggestions above? Also, I wonder if this is related to https://github.com/numba/numba/pull/2950 and it's actually the alignment that's throwing the dispatcher off.

**Comment by stuartarchibald** [comment]
@jzwinck thanks for getting back to us on this. What you describe is suspiciously similar to https://github.com/numba/numba/issues/2937 and the WIP fix for it in #2950 (if you are confident building Numba from source, perhaps patch in the change to `numba/_typeof.c` and give it a go?).

Might also be worth looking at the `ndarray.flags` for your mmap'd data to see if any of it has `ALIGNED: False`, which I think could possibly trigger a similar situation to #2937 under certain signature registration sequences.

You might find `numba.typeof()` useful for seeing what Numba thinks your data types are. Also, if you want to create explicit signatures then `numba.types` is a quick way to do this, e.g.
```
In [39]: numba.types.float64[1::, ::1]
Out[39]: array(float64, 2d, C)

In [40]: numba.types.float64[::1, ::1]
Out[40]: array(float64, 2d, F)

In [41]: numba.types.float64[1::, 1::]
Out[41]: array(float64, 2d, A)
```
signatures can be composed of these objects or just written as strings, like:
```
@njit('float64[1::, 1::](float64[1::, 1::],float64[1::, 1::])')
def foo(a, b):
    return a + b
```
gives `foo.signatures` : `[(array(float64, 2d, A), array(float64, 2d, A))]`.

If you can create a reproducer for the problem you are seeing then please share it and we'll take a look. Thanks.

**Comment by jiangyue12392** [comment]
The code below reproduces the bug mentioned above:

```python
import numpy as np
import numba

@numba.njit(nogil=True)
def std_2d(array, mean):
    """Input array is 2D, plus column-wise mean (1D)
    Returns column-wise standard deviation.
    Same as array.std(0)
    """
    assert array.ndim == 2
    length, width = array.shape

    variance = np.zeros(width)  # result is always float

    for row in range(length):
        for col in range(width):
            diff = array[row, col] - mean[col]
            variance[col] += diff * diff

    return np.sqrt(variance / length)


# Generate a 100 x 20 numpy array where each column is assigned as
# a float of column number x 1000 (1000 is there only to amplify
# the difference)
gen_arr = np.empty((100, 20))
for col in range(20):
    gen_arr[:, col] = np.full(100, col * 1000)

# Save the generated array into file in fortran order
gen_arr = np.asfortranarray(gen_arr)
np.save('arr_file', gen_arr)

times = 10
for rd in range(times):
    # Load array from file with mmap_mode='r'
    loaded_arr = np.load('arr_file.npy', mmap_mode='r')

    if (rd % 2 == 0):
        # Note that slicing by full column maintains fortran order of the data
        arr_to_cal  = loaded_arr[:, 10:12]
    else:
        # Slicing a portion of the data in the way below creates a c order array
        # due to some implicit mechanisms
        arr_to_cal  = loaded_arr[10:90, 10:12]

    mean = np.mean(arr_to_cal, axis=0) # Column-wise mean as required by std_2d function
    expected = arr_to_cal.std(0) # This should give 0 as the whole column value is the same
    actual = std_2d(arr_to_cal, mean)

    # This prints the storage order of the array
    print('Round:', rd, ' Array Order:','Fortran' if np.isfortran(arr_to_cal) else 'C')

    # Reports error if actual sd calculated from std_2d is non-zero
    assert (actual < 0.001).all()
```
On numba version 0.40 it still produces assertion error.
However, the problem does not persist for numba version 0.41 onwards.

**Comment by jzwinck** [comment]
Fixed in 0.41, most likely by #2950.

**Comment by stuartarchibald** [comment]
@jzwinck Thanks for confirming the fix, much appreciated.
